### ISSUE TITLE
Subscription request fails to complete returned Task when a slave is disconnected

### ISSUE THREAD

**Issue by michaelgrosner** [StackExchange/StackExchange.Redis#349]
Hello,

I am able to reliably reproduce an issue where the Task object returned by `ISubscriber.SubscribeAsync` never completes. The issue appears when the given master instance is connected and another Redis instance provided to the ConfigurationOptions is disconnected. I am using version 1.0.488 of StackExchange.Redis. 

Given the following code:

``` C#
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using StackExchange.Redis;

namespace RedisSubscriberTest
{
    class Program
    {
        private static ConfigurationOptions _config;
        private static ConnectionMultiplexer _mux;
        private static ISubscriber _sub;

        private const string KeyPrefix = "allkeysfortest__longname____";
        private const string KeyCache = KeyPrefix + ":keyCache";

        public static void Main(string[] argv)
        {
            Task.Run(() => MainAsync(argv)).Wait();
        }

        private static async Task MainAsync(string[] argv)
        {
            _config = ConfigurationOptions.Parse(argv.First());
            _config.DefaultDatabase = 1;
            _config.AbortOnConnectFail = false;

            _mux = ConnectionMultiplexer.Connect(_config);
            _mux.PreserveAsyncOrder = true;

            _sub = _mux.GetSubscriber();

            Console.WriteLine("Begin Test {0} ... Connection Statuses: {1}", argv.Last(),
                string.Join(" ",
                    _mux.GetEndPoints().Select(ep => $"{ep}={_mux.GetServer(ep).IsConnected}")));

            var start = DateTime.UtcNow;

            if (argv.Last() == "populate")
            {
                await PopulateAsync();
            }
            else
            {
                await RunTestAsync(argv);
            }

            var t = DateTime.UtcNow - start;
            Console.WriteLine("End Test, took {0}", t.TotalMilliseconds);

            await Task.Delay(TimeSpan.FromSeconds(60000));
        }

        private static async Task PopulateAsync()
        {
            var generateRandomData =
                Enumerable.Range(0, 4500)
                    .Select(x => $"{KeyPrefix}:{x}")
                    .Select(key => (RedisValue) key)
                    .ToArray();

            await _mux.GetDatabase().SetAddAsync(KeyCache, generateRandomData);
        }

        private static async Task RunTestAsync(string[] argv)
        {
            var keys = (await (argv.Last() == "sscan" ? KeysScan() : KeyMembers())).ToArray();

            Console.WriteLine("Downloaded {0} keys", keys.Length);

            await Task.WhenAll(keys.SelectMany(SubscribeForKey));
        }

        private static Task<IEnumerable<RedisValue>> KeyMembers()
        {
            return _mux.GetDatabase().SetMembersAsync(KeyCache).ContinueWith(t => t.Result as IEnumerable<RedisValue>);
        }

        private static Task<IEnumerable<RedisValue>> KeysScan()
        {
            return Task.FromResult(_mux.GetDatabase().SetScan(KeyCache, pageSize: 1000));
        }

        private static IEnumerable<Task> SubscribeForKey(RedisValue key)
        {
            yield return SubscribeFor($"upd:{key}");
            yield return SubscribeFor($"del:{key}");
        }

        private static async Task SubscribeFor(string topic)
        {
            var subTask = _sub.SubscribeAsync(topic,
                (channel, value) => Console.WriteLine("Got data on {0} {1}", channel, value));

            if (await Task.WhenAny(Task.Delay(TimeSpan.FromSeconds(10)), subTask) == subTask)
                return;

            Console.WriteLine("Task for {0} timed out, endpoint: {1}", topic, _sub.IdentifyEndpoint(topic));
        }
    }
}
```

This example program can first populate in Redis all of the subscription topics needed for the test by using the "populate" command-line switch. Next, to run the test, (the dev-redis instance is the master, localhost redis is a downed slave):

```
C:\Users\grosnerm\Desktop\Logs\RedisSubscriberTest\RedisSubscriberTest\bin\Debug>RedisSubscriberTest.exe "dev-redis:6385,localhost:6388,SyncTimeout=5000" sscan
Begin Test sscan ... Connection Statuses: Unspecified/dev-redis:6385=True Unspecified/localhost:6388=False
Task for upd:allkeysfortest__longname____:3532 timed out, endpoint: Unspecified/dev-redis:6385
Task for del:allkeysfortest__longname____:2114 timed out, endpoint: Unspecified/dev-redis:6385
Task for del:allkeysfortest__longname____:920 timed out, endpoint: Unspecified/dev-redis:6385
End Test, took 10028.3927
```

In this case, 3 of the 4500 subscription requests fail to complete within 10 seconds (all the others complete in single-digit milliseconds).  Curiously, every time I reproduce this issue the "2114" subscription request fails.

If I use the redis-cli tool, I can publish to my test app:

`127.0.0.1:6385> PUBLISH "upd:allkeysfortest__longname____:2114" something`

And the app will correctly handle the publication:

`Got data on upd:allkeysfortest__longname____:2114 something`

This means the underlying subscription in SubscribeAsync actually succeeded, however, the Task returned by it never actually runs to completion. 

I'm only able to reproduce this when the master is connected but not any of the other provided instances. This issue does not manifest when both master/slave are up, when only the master is provided and that's up, slave is up but master is down. Please let me know if you need any more information on this issue.

Michael

**Comment by NickCraver** [comment]
Curious...I think this relates to the master-only status checking issues elsewhere, we'll take a look soon as time allows.

**Comment by piccit** [comment]
I'm seeing the same behavior with Set tasks (on version 1.0.488 on .Net 4.5.1 and also on 1.0.333 on .Net 4.0). Here's a straightforward way to reproduce: 


`static void Main(string[] args)
        {
            var opts = ConfigurationOptions.Parse("localhost:6379,localhost:6380");
            var connMulti = ConnectionMultiplexer.Connect(opts);
            connMulti.PreserveAsyncOrder = false;
            Console.WriteLine("Connected: " + connMulti.IsConnected);

            //show that sync works without issue
            for (int i = 0; i < 20; i++)
            {
                string key = "sync" + i.ToString();
                Console.WriteLine("Sync set: " + connMulti.GetDatabase().StringSet(key, key) );
            }

            List<Task<bool>> writeTasks = new List<Task<bool>>();
            for (int i = 0; i < 20; i++)
            {
                string key = "foo" + i.ToString();
                var t = connMulti.GetDatabase().StringSetAsync(key, key);
                writeTasks.Add(t);
            }

            if (!Task.WaitAll(writeTasks.ToArray(), TimeSpan.FromSeconds(10)))
            {
                Console.BackgroundColor = ConsoleColor.DarkRed;
                Console.WriteLine("Tasks timed out!");
                Console.ResetColor();
            }

            foreach (var t in writeTasks)
            {
                Console.WriteLine("\tTask status: " + t.Status + " Result: " + 
                    (t.Status == TaskStatus.RanToCompletion ? t.Result.ToString() : "null"));
            }
        }`

Which will give the following output when one of the servers is not running:
`Connected: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Sync set: True
Tasks timed out!
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: WaitingForActivation Result: null
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True
        Task status: RanToCompletion Result: True`

But all of the keys are successfully written to Redis.
 
In poking around in the debugger, it looks like for the each task that never completes, a ResultBox tries to set the task completion source result to False. I didn't have a chance to look into where the false is coming from (since all the keys get written), but as far as I can guess, it's somehow setting the wrong task completion source when it is in this case.

Setting preserver async order did not appear to make any difference.

I even played around with trying to recover from this situation by disposing the multiplexer and creating a new one once all of the servers are online -- but this didn't seem to help at all. The only way I could recover was by restarting the application when all of the servers were online.

I also found that the same issue can arise after a Redis server failover.

**Comment by piccit** [comment]
I was **unable** to reproduce using 1.1.572-alpha (still on .net 4.5.1). In testing the alpha versions, it looks like the issue was fixed with 1.1.553-alpha. Is there a way to determine what changes where made from 1.1.0-alpha2 to 1.1.553-alpha? It would be nice to understand the cause/fix (plus it's easier for us in the short run to patch the version we have now than it is to update to a whole new version)

**Comment by piccit** [comment]
In looking at the history of what has changed since 1.1.0-alpha2, I think the fix comes from commit: [a2328ef](https://github.com/StackExchange/StackExchange.Redis/commit/a2328ef84c6997de77be91b79afd2dc4bce9655f) 

I tested patching just that change into 1.0.333 and I can no longer re-create the issue. I will test with that change in place more today - but it's looking promising

**Comment by piccit** [comment]
So that change does fix the problem when creating the multiplexer with a server down. However, the problem still sometimes happens after a failover (it doesn't happen every time, which makes it hard to reproduce).

As for not being able to recover with a dispose, I'm wondering if maybe something that is static isn't getting cleaned up?

**Comment by piccit** [comment]
I've added lots of extra debug logging and from that I believe the issue is that the same `ResultBox `ends up in the store more than once (causing the same box to be used for 2+ messages, causing the the `TaskCompletionSource `from the first message to be lost).

I believe the `ResultBox `ends up in the store because it looks like `TryComplete `can get called on the same PING message twice in a disconnect scenario.

In PhysicalBridge `OnDisconnected` has this:
```cs
            // if the next thing in the pipe is a PING, we can tell it that we failed (this really helps spot doomed connects)
            // note that for simplicity we haven't removed it from the queue; that's OK
            int count;
            var ping = queue.PeekPing(out count);
            if (ping != null)
            {
                Trace("Marking PING as failed (queue length: " + count + ")");
                ping.Fail(failureType, null);
                CompleteSyncOrAsync(ping);
            }
            oldState = default(State); // only defined when isCurrent = true
            if (isCurrent = (physical == connection))
            {
                Trace("Bridge noting disconnect from active connection" + (isDisposed ? " (disposed)" : ""));
                oldState = ChangeState(State.Disconnected);
```

Which is going to `TryComplete` (via `CompleteSyncOrAsync(ping)`) on the PING (if there is one). But then it's going to call `ChangeState `which will call `AbortUnsent` which is _also_ going to call `TryComplete `on the PING(if there was one) since it wasn't removed from the queue.

Somewhere in there must be a race condition, since it doesn't appear to happen every time (or maybe the store is just full sometimes when `TryComplete `is calling `UnwrapAndRecycle `on the second `ResultBox`


Am I missing something, @mgravell or @NickCraver, or does PeekPing actually need to dequeue the ping - or `AbortUnsent `could ignore `dead[0]` if it's a PING

**Comment by mgravell** [comment]
Ouch. That looks awkward. Will check.

**Comment by piccit** [comment]
And now that I think about my comment about commit a2328ef, I see why it's a race condition.

With that commit we now null out the `resultBox` after calling `TryComplete` in `Message` when we know it will have already called `UnwrapAndRecycle`. Which mostly fixes the issue of calling `TryComplete` more than once on the same Message (which is why with that change it's very difficult to reproduce the issue).

However, that block of code in Message is a race condition, since two threads could potentially get into `TryComplete `at roughly the same time (due to the peek Ping thing).

So I'm not sure if changing `PeekPing `is the way to go or if there should be something more thread-safe in Message `TryComplete`

**Comment by mgravell** [comment]
The more I think about it, the more I conclude we simply need to lose those
reused tasks. Simply not worth it, and the small number of scenarios where
they are safe is probably too expensive (both in runtime CPU cycles and
dev-time reasoning) to reliably detect and maintain.

**Comment by mgravell** [comment]
/s/tasks/messages

**Comment by mgravell** [comment]
I think I see some dubious ways this could happen - have tried to address in 4fcb1269ea583132ce2704ce63be83cc91073f04, but this is a complicated one. If result-boxes are getting added to the store incorrectly.... yeah, bad things.

**Comment by piccit** [comment]
I think to make the `Message` class fool proof (as far as calling `TryComplete` on it from multiple threads), it'd have to be something like this:

```cs
        public bool TryComplete(bool isAsync)
        {
			var currBox = Interlocked.Exchange(ref resultBox, null);
            if (currBox != null)
            {
                var ret = currBox.TryComplete(isAsync);

                if (!(ret && isAsync))
                {
                    //put result box back if it was not already recycled
					Interlocked.Exchange(ref resultBox, currBox);
                }
```

But I'm not sure if that would have negative performance impacts.

**Comment by mgravell** [comment]
That indeed should work - PR welcome! Do you see a code path that would lead to TryComplete being called twice for the same message? (with the exception of a few of the fire+forget messages that are shared - they never have a result-box)

**Comment by piccit** [comment]
Just the Ping path I described above in a disconnect scenario. I think the solution there is to actually dequeue the ping in `PeekPing`. I'll put together a PR, I just have to get git setup again on my machine

**Comment by mgravell** [comment]
I'll see what I can repro on that ping path, thanks for the heads-up. That
would be a big "d'oh" moment, so if you're right (and you probably are),
this is very much in need of fixing.

**Comment by piccit** [comment]
PR #376 submitted

**Comment by NickCraver** [comment]
Fixed in #376
